mypy/newsemanal/semanal.py

```diff
@@ -1726,13 +1765,23 @@ def anal_type(self, t: Type, *,
                   allow_tuple_literal: bool = False,
                   allow_unbound_tvars: bool = False,
                   report_invalid_types: bool = True,
-                  third_pass: bool = False) -> Type:
+                  third_pass: bool = False) -> Optional[Type]:
+        """Semantically analyze a type.
+
+        Return None only if some part of the type couldn't be bound *and* it referred
+        to an incomplete namespace. In case of other errors, report an error message
+        and return AnyType.
+        """
         a = self.type_analyzer(tvar_scope=tvar_scope,
                                allow_unbound_tvars=allow_unbound_tvars,
                                allow_tuple_literal=allow_tuple_literal,
                                report_invalid_types=report_invalid_types,
                                third_pass=third_pass)
+        prev_incomplete = self.num_incomplete_refs
         typ = t.accept(a)
+        if self.num_incomplete_refs != prev_incomplete:
```

Comment:
This way looks a bit tricky, maybe we can get this somehow from the return type (or use an exception)? This is fine for a follow-up PR if you agree.

Reply:
Using return type may be tricky, since this needs to also work with the main AST visitor, which doesn't return anything. Using an exception may be a good idea. Let's revisit this later once we have a more complete implementation.